### PR TITLE
remove ecdhRaw usage and replace it with ecdh+custom hash function

### LINKED ISSUE #222: Use custom hash function for eth1 ecdh
We currently extend upstream libsecp256k1 to add a `secp256k1_ecdh_raw` function - since this was implemented, upstream has added the ability do do custom hash functions that could be used to implement this ethereum-specific way of deriving keys - as such, we can stop patching upstream and move the logic into `nim-eth`.

Upstream discussion:
https://github.com/bitcoin-core/secp256k1/issues/352

Implementation blueprint:
https://github.com/bitcoin-core/secp256k1/blob/master/src/modules/ecdh/tests_impl.h#L18-L25

### PR DESCRIPTION
**Pull request by jangko**
fix #222

### CODE REVIEW

**Review by arnetheduck — APPROVED** [review]

File: eth/keys.nim [hunk 1]
```diff
@@ -17,15 +17,17 @@
...
+  stew/[byteutils, objects, results, ptrops],
   ./common/eth_hash
 
 from nimcrypto/utils import burnMem
 
 export secp256k1, results, rand
 
 const
-  KeyLength* = SkEcdhRawSecretSize - 1
+  FullKeyLength* = 33
```
Comment (cheatfate) [inline eth/keys.nim:28]: Where this number `33` come from?

File: eth/keys.nim [hunk 2]
```diff
@@ -230,11 +234,28 @@ func verify*(sig: SignatureNR, msg: openArray[byte], key: PublicKey): bool =
...
-  copyMem(addr result.data[0], unsafeAddr(tmp.data[1]), sizeof(result))
-
-func ecdhRawFull*(seckey: PrivateKey, pubkey: PublicKey): SharedSecretFull =
-  SharedSecretFull(ecdhRaw(SkSecretKey(seckey), SkPublicKey(pubkey)))
+proc ecdhSecretHash(output: ptr byte, x32, y32: ptr byte, data: pointer): cint
+                    {.cdecl, raises: [].} =
+  copyMem(output, x32, 32)
+  return 1
+
+func ecdhSecret*(seckey: PrivateKey, pubkey: PublicKey): SharedSecret =
```
Comment (cheatfate) [inline eth/keys.nim:242]: This is really bad name `ecdhSecret`, it should be `ecdhSharedSecret` at least.

File: eth/keys.nim [hunk 3]
```diff
@@ -230,11 +234,28 @@ func verify*(sig: SignatureNR, msg: openArray[byte], key: PublicKey): bool =
...
-  let tmp = ecdhRaw(SkSecretKey(seckey), SkPublicKey(pubkey))
-
-  # Remove first byte!
-  copyMem(addr result.data[0], unsafeAddr(tmp.data[1]), sizeof(result))
-
-func ecdhRawFull*(seckey: PrivateKey, pubkey: PublicKey): SharedSecretFull =
-  SharedSecretFull(ecdhRaw(SkSecretKey(seckey), SkPublicKey(pubkey)))
+proc ecdhSecretHash(output: ptr byte, x32, y32: ptr byte, data: pointer): cint
+                    {.cdecl, raises: [].} =
+  copyMem(output, x32, 32)
```
Comment (cheatfate) [inline eth/keys.nim:239]: Where this number `32` come from?

**Review by cheatfate — CHANGES_REQUESTED** [review]

File: eth/keys.nim (hunk 1 above)
Comment (jangko) [reply eth/keys.nim]: fixed

File: eth/keys.nim (hunk 2 above)
Comment (jangko) [reply eth/keys.nim]: changed to `ecdhSharedSecret`

File: eth/keys.nim (hunk 3 above)
Comment (jangko) [reply eth/keys.nim]: replaced with `KeyLength` constant

**Review by cheatfate — APPROVED** [review]

**Pull request merged**